### PR TITLE
feature: get flow run logs from CLI

### LINKED ISSUE #7814: Get flow's logs using command line
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar request and didn't find it.
- [X] I searched the Prefect documentation for this feature.

### Prefect Version

2.x

### Describe the current behavior

Based on the below conversation,
https://prefect-community.slack.com/archives/CL09KU1K7/p1670480537109379

It would be good if we get flow logs with command line

### Describe the proposed behavior

 In prefect1 we can use the `prefect get` command to get logs. expecting similar feature in prefect2 as well.

### Example Use

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ohadch**
Usage: prefect flow-run logs <flow-run-id>
closes #7814

This PR attempts to resolve issue #7814 .

It adds a new `logs` command to the `flow-runs` command in the CLI.

### Example

Usage is as follows:

```
prefect flow-run logs <flow-run-id>
```

The CLI responds as follows:

```
2022-12-23T14:44:16.847706+00:00  | INFO  | Finished in state Completed()
```

This is my first PR here. Waiting for your review and opinions :)

Ohad

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes tests or only affects documentation.
- [ ] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`

### CODE REVIEW

**Comment by serinamarie** [comment]
Hi @ohadch, thanks so much for contributing to Prefect! When reading the logs we would want to handle paginated queries as well, something like 

```python
offset = 0
limit = 200
more_logs = True
all_logs = []

while more_logs:

    logs = await client.read_logs(
        log_filter=log_filter, limit=limit, offset=offset
        )
    if len(logs) > 0:
        all_logs.extend(logs)
        offset += limit
    else:
        more_logs = False
```
because otherwise we will be limited to 200 logs.

**Comment by madkinsz** [comment]
We'll also probably want to display the logs after each paginated query rather than pushing them all to a list and displaying it in one go.

**Comment by ohadch** [comment]
Hey @serinamarie and @madkinsz ,

Thanks for welcoming me and for reviewing my code. I implemented pagination as requested, the output now looks as follows:

```
2022-12-24T07:56:07.243937+00:00  | INFO  | Log 48 from flow_run d6b76008-406c-4f69-a7b7-d1186e47a94d.
2022-12-24T07:56:07.243958+00:00  | INFO  | Log 49 from flow_run d6b76008-406c-4f69-a7b7-d1186e47a94d.
Press enter to see more logs...
2022-12-24T07:56:07.243982+00:00  | INFO  | Log 50 from flow_run d6b76008-406c-4f69-a7b7-d1186e47a94d.
2022-12-24T07:56:07.244005+00:00  | INFO  | Log 51 from flow_run d6b76008-406c-4f69-a7b7-d1186e47a94d.
```

I would like to hear your opinion about the print format. I got the general idea from the way that the logs are printed during the live flow run, but I did not find where the format is declared. Can you maybe either guide me where to find it, or provide me with the desired spec?

Thanks again!
Ohad

**Comment by madkinsz** [comment]
The [logging module](https://github.com/PrefectHQ/prefect/tree/main/src/prefect/logging) may be helpful. The format is defined in the [logging YAML file](https://github.com/PrefectHQ/prefect/blob/main/src/prefect/logging/logging.yml#L16-L21). You may be able to construct a [flow run logger](https://github.com/PrefectHQ/prefect/blob/main/src/prefect/logging/loggers.py#L58) but you'd want to be careful to disable sending those logs to the server ([example](https://github.com/PrefectHQ/prefect/blob/main/src/prefect/engine.py#L706)). It might be more trouble than it's worth to use the actually logger. Maybe you can use a logging formatter without using a logger itself?

**Comment by madkinsz** [comment]
I like the idea of requiring user input to go to the next page, but perhaps we should put that behind a `-i/--interactive` flag and just show all of the logs by default.

**Comment by ohadch** [comment]
@madkinsz 
1. Regarding the logger - I think we should not use the logger there, as we just intend to show logs and not to create new ones. I only want to use the formatter or construct the string according to it in order to keep the UX consistent.
2. I thought you wanted the user input to exist, whoops :) Actually I think this behavior may be counter intuitive and a bit redundant, so maybe we should avoid adding both the flag and the keypress, and just load (and print) the pages each at a time as you suggested?

**Comment by madkinsz** [comment]
@ohadch that's fine with me

File: tests/cli/test_flow_run.py [hunk 1]
```diff
@@ -275,3 +277,98 @@ def test_wrong_id_exits_with_error(self):
...
+            ],
+            expected_code=0,
+            expected_output_contains=[
+                f"Flow run '{flow_run.name}' - Log {i} from flow_run {flow_run.id}."
+                for i in range(self.PAGE_SIZE - 1)
+            ],
+        )
+
+    @pytest.mark.parametrize("state", [Completed, Failed, Crashed, Cancelled])
+    async def test_when_more_than_page_size_logs_exists_then_pagination(
```
Comment (serinamarie) [inline tests/cli/test_flow_run.py:323]: Personally, I think that something like `test_when_num_logs_greater_than_page_size_then_pagination` might be clearer.

File: tests/cli/test_flow_run.py [hunk 2]
```diff
@@ -275,3 +277,98 @@ def test_wrong_id_exits_with_error(self):
...
+        self, orion_client, state
+    ):
+        # Given
+        flow_run = await orion_client.create_flow_run(
+            name="scheduled_flow_run", flow=hello_flow, state=state()
+        )
+
+        # Create more than page size flow run logs
```
Comment (serinamarie) [inline tests/cli/test_flow_run.py:331]: Like above, something like "Create enough flow run logs to result in pagination (page_size > 200)" might be clearer.

File: tests/cli/test_flow_run.py [hunk 3]
```diff
@@ -275,3 +277,98 @@ def test_wrong_id_exits_with_error(self):
...
+            invoke_and_assert,
+            command=[
+                "flow-run",
+                "logs",
+                str(flow_run.id),
+            ],
+            expected_code=0,
+            expected_output_contains=[
+                f"Flow run '{flow_run.name}' - Log {i} from flow_run {flow_run.id}."
+                for i in range(self.PAGE_SIZE)
```
Comment (serinamarie) [inline tests/cli/test_flow_run.py:355]: If I'm understanding this test correctly, the goal here is to create 201 logs and then return those 201 logs. Would we want this to be `self.PAGE_SIZE + 1` to match with the number of created logs? Regardless, adding `+ 1` still results in a successful test.

**Review by serinamarie — COMMENTED** [review]
Hi @ohadch, thanks for adding in pagination handling! I left a few nitpick suggestions.

File: tests/cli/test_flow_run.py (hunk 3 above)
Comment (ohadch) [reply tests/cli/test_flow_run.py:355]: That's a good catch! Indeed `self.PAGE_SIZE + 1` is the desired case.

File: tests/cli/test_flow_run.py (hunk 1 above)
Comment (ohadch) [reply tests/cli/test_flow_run.py:323]: Sure 🙏

File: tests/cli/test_flow_run.py (hunk 2 above)
Comment (ohadch) [reply tests/cli/test_flow_run.py:331]: Like above - I agree.

**Comment by ohadch** [comment]
Hey @serinamarie ,

Thank you very much for your review. I fixed the code according to your suggestions.

Happy new year!

File: src/prefect/cli/flow_run.py [hunk 4]
```diff
@@ -135,3 +136,42 @@ async def cancel(id: UUID):
...
+
+            # Print the logs
+            for log in page_logs:
+                app.console.print(
+                    # Print following the flow run format (declared in logging.yml)
+                    f"{pendulum.instance(log.timestamp).to_datetime_string()}.{log.timestamp.microsecond // 1000:03d} | {logging.getLevelName(log.level):7s} | Flow run {flow_run.name!r} - {log.message}",
+                    soft_wrap=True,
+                )
+
+            # Wait for the user to press enter to get the next page
```
Comment (serinamarie) [inline src/prefect/cli/flow_run.py:172]: This PR no longer includes user input, though, correct?

Comment (ohadch) [reply src/prefect/cli/flow_run.py:172]: Correct. As discussed with @madkinsz , we omitted it, and we only load and print the pages chunk by chunk.

Comment (serinamarie) [reply src/prefect/cli/flow_run.py:172]: I would think we don't need that comment on line 172 then

Comment (serinamarie) [reply src/prefect/cli/flow_run.py:172]:
```suggestion
```

**Comment by ohadch** [comment]
That's correct, fixed. Thanks!

**Review by serinamarie — APPROVED** [review]
@ohadch This looks good - thanks so much for your contribution 🙂

**Comment by ohadch** [comment]
I had the pleasure, thank you @serinamarie !

**Comment by serinamarie** [comment]
Related to this, I have linked to some additional enhancements adding onto this, should you feel so inclined 😀 

Have a happy new year! 🎉

**Comment by ohadch** [comment]
@serinamarie Sure, I already commented on the `--head` one so either of you assign me to it.

Have a happy new year as well!

**Comment by serinamarie** [comment]
@ohadch Amazing, thank you!

**Pull request merged**